targets/TARGET_STM/serial_api.c

```diff
@@ -661,6 +661,164 @@ HAL_StatusTypeDef init_uart(serial_t *obj)
     return HAL_UART_Init(huart);
 }
 
+UART_HandleTypeDef * dmauart;
+
+void DMA1_Channel5_IRQHandler(void)
+{
+    HAL_DMA_IRQHandler(dmauart->hdmarx);
+}
+
+void DMA1_Channel4_IRQHandler(void)
+{
+    HAL_DMA_IRQHandler(dmauart->hdmatx);
+}
+
+/**
+  * @brief  Rx Transfer completed callback
+  * @param  UartHandle: UART handle
+  * @note   This example shows a simple way to report end of DMA Rx transfer, and
+  *         you can add your own implementation.
+  * @retval None
+ */
+
+typedef void(*DmaCompleteCallback)(uint32_t);
+
+DmaCompleteCallback callback_dma_rx;
+uint32_t callback_id_dma_rx;
+
+DmaCompleteCallback callback_dma_tx;
+uint32_t callback_id_dma_tx;
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef* UartHandle)
+{
+    /* Set transmission flag: transfer complete*/
+    serial_idle_callback();
+}
+
+void HAL_UART_TxCpltCallback(UART_HandleTypeDef* UartHandle)
+{
+    if (callback_dma_tx) {
+        callback_dma_tx(callback_id_dma_tx);
+    }
+}
+
+void serial_idle_callback(void)
+{
+    if (callback_dma_rx) {
+        callback_dma_rx(callback_id_dma_rx);
+    }
+}
+
+static void uart1_irq_idle(void)
+{
+    int8_t id = get_uart_index(UART_1);
+
+    if (id >= 0) {
+        UART_HandleTypeDef* huart = &uart_handlers[id];
+
+        if (__HAL_UART_GET_FLAG(huart, UART_FLAG_IDLE) != RESET) {
+            if (__HAL_UART_GET_IT(huart, UART_IT_IDLE) != RESET) {
+                __HAL_UART_CLEAR_FLAG(huart, UART_FLAG_IDLE);
+                serial_idle_callback();
+            }
+        }
+
+        if (__HAL_UART_GET_FLAG(huart, UART_FLAG_TC) != RESET) {
+            if(__HAL_UART_GET_IT(huart, UART_IT_TC) != RESET) {
+                __HAL_UART_CLEAR_FLAG(huart, UART_FLAG_TC);
+                huart->gState = HAL_UART_STATE_READY;
+            }
+        }
+    }
+}
+
+void serial_rx_dma_init(serial_t *obj, DMA_HandleTypeDef *hdma_rx, uint8_t* buffer, size_t buffer_size, void(*DmaCompleteCallback)(uint32_t), uint32_t id)
```

Comment:
```suggestion
void serial_rx_dma_init(serial_t *obj, DMA_HandleTypeDef *hdma_rx, uint8_t* buffer, size_t buffer_size, DmaCompleteCallback cb, uint32_t id)
```
and then in line `callback_dma_rx = cb;`